e2store/src/era2.rs

```diff
@@ -0,0 +1,313 @@
+        }
+        let account = AccountEntry::try_from(&self.e2store_stream.next_entry()?)?;
+        self.storage_entries_left_to_decode = account.storage_count;
+        Ok(Era2StreamEntry::Account(account))
+    }
+
+    pub fn append_entry(&mut self, entry: &Era2StreamEntry) -> anyhow::Result<usize> {
+        let size = match entry {
+            Era2StreamEntry::Account(account) => {
+                if self.storage_entries_left_to_decode != 0 {
+                    panic!("Invalid append entry state: expected a storage entry, got an account entry. Still have {} storage entries left to append", self.storage_entries_left_to_decode);
+                }
+
+                self.storage_entries_left_to_decode = account.storage_count;
+                let entry: Entry = account.clone().try_into()?;
+                self.e2store_stream.append_entry(&entry)?;
+                entry.value.len()
+            }
+            Era2StreamEntry::Storage(storage) => {
+                if self.storage_entries_left_to_decode == 0 {
+                    panic!("Invalid append entry state: expected an account entry, got a storage entry. No storage entries left to append for the account");
+                }
+
+                self.storage_entries_left_to_decode -= 1;
+                let entry: Entry = storage.clone().try_into()?;
+                self.e2store_stream.append_entry(&entry)?;
+                entry.value.len()
+            }
+        };
+        Ok(size)
+    }
+}
+
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub enum Era2StreamEntry {
+    Account(AccountEntry),
+    Storage(StorageEntry),
+}
+
+#[derive(Clone, Eq, PartialEq, Debug, RlpEncodable, RlpDecodable)]
+pub struct AccountEntry {
+    pub address_hash: B256,
+    pub account_state: AccountState,
+    pub bytecode: Vec<u8>,
+    pub storage_count: u16,
+}
+
+impl TryFrom<&Entry> for AccountEntry {
+    type Error = anyhow::Error;
+
+    fn try_from(entry: &Entry) -> Result<Self, Self::Error> {
+        ensure!(
+            entry.header.type_ == 0x08,
+            "invalid account entry: incorrect account type"
+        );
+        ensure!(
+            entry.header.reserved == 0,
+            "invalid account entry: incorrect account reserved bytes"
+        );
+        let mut decoder = snap::read::FrameDecoder::new(&entry.value[..]);
+        let mut buf: Vec<u8> = vec![];
+        decoder.read_to_end(&mut buf)?;
+        let account = Decodable::decode(&mut buf.as_slice())?;
+        Ok(account)
+    }
+}
+
+impl TryInto<Entry> for AccountEntry {
+    type Error = anyhow::Error;
+
+    fn try_into(self) -> anyhow::Result<Entry> {
+        let rlp_encoded = alloy_rlp::encode(self);
+        let buf: Vec<u8> = vec![];
+        let mut encoder = snap::write::FrameEncoder::new(buf);
+        let _ = encoder.write(&rlp_encoded)?;
+        let encoded = encoder.into_inner()?;
+        Ok(Entry::new(0x08, encoded))
+    }
+}
+
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct StorageEntry {
```

Comment:
nit: maybe make this `pub struct StorageEntry(Vec<StorageItem>);`? (in which case, I would implement `Deref` trait)

If not, I would at least rename field to `items`?

Reply:
This is cool